### ISSUE TITLE
[0.7] Suspend inner runs twice, causing double-render of returned View

### ISSUE THREAD

**Issue by BakerNet** [leptos-rs/leptos#2956]
**Describe the bug**

When a Suspend is created to await a resource, the inner future is run twice.  This results in the returned Views rendering twice.    

In my case, this was leading to duplicated websocket connections, probably because of a race condition between web_sys callbacks and Leptos cleanup.

**Leptos Dependencies**

0.7 (only tested on `-beta5` and `fix-refetch`)

**To Reproduce**

See recreation in this start-axum-0.7 fork branch:
https://github.com/BakerNet/start-axum-0.7/tree/demo/suspend-bug

https://github.com/BakerNet/start-axum-0.7/blob/demo/suspend-bug/src/app.rs#L75

**Expected behavior**

The Suspend inner only runs once inside a Suspense/Transition block unless refetched or source data changes

**Screenshots**

https://github.com/user-attachments/assets/f92c1460-1024-429c-850a-290e2c271160

**Comment by gbj** [comment]
AFAICT this is working as designed.

In order to register both synchronous and async Resource reads, `Suspense` will walk over the tree of its children once. It will poll any `Suspend`, but not wait for it: i.e., if there is an `.await` point that is not ready yet, it will not progress beyond that point.

When rendering, the `Future` in a `Suspend` is polled once to check whether it is immediately, synchronously available. If it is, then its contents are simply rendered, and if not, then the `Future` is spawned.

In the example you provide,
1) the "side efffect" (updating the signal) is done in the synchronous part of the `Suspend`, before the `.await` point, and
2) the server function/resource resolves synchronously on the server

As a result, the `Suspend` is always ready the first time it is polled on the server, so the effectful synchronous code ends up executing twice.

If the server function is actually async on the server (i.e., it is not immediately resolved when called) and the side effect is after the `.await`, you get the result you're anticipating

```rust
#[server]
pub async fn get_thing(thing: ()) -> Result<(), ServerFnError> {
    tokio::time::sleep(std::time::Duration::from_millis(25)).await;
    Ok(thing)
}

// ...

                Suspend::new(async move {
                    let _ = resource.await;
                    count.update_untracked(|x| *x += 1);
```

Another workaround here is not using `Suspense`, but just the `Suspend` (this is allowed now and doesn't have the same double-polling effect, because it doesn't need to check for synchronous resource reads).

I'm happy to try to help debug something closer to your real example with the websockets if needed. I don't think that the behavior will change here, as it's pretty fundamental to the async/sync Suspense system, but if there's an example that isn't working as intended I'm happy to fix it.

**Comment by metatoaster** [comment]
> If the server function is actually async on the server (i.e., it is not immediately resolved when called) and the side effect is after the `.await`, you get the result you're anticipating

Since this was something I've observed while building the test case for #2937 I took the liberty to test this again with the additional modification to [the same gist](https://gist.github.com/metatoaster/2798451bd9bc99d57b73c46d7cc8ca9a) I've used as the test bed, the code after await will still be called twice _under CSR_ as the first request will be held up by the unresolved future.  I've used logging instead to more clearly illustrate the flow that the futures have taken.  The relevant logging/action parts of closure that returns the `Suspend::new` for the component looks like this when inlined with irrelevant bits removed:

```rust
    view! {         
        <Suspense>
            {move || {
                leptos::logging::log!("inspect_view closure invoked");
                Suspend::new(async move {
                    leptos::logging::log!("inspect_view Suspend::new() called");
                    let result = res_inspect.await.map(|item| {
                        leptos::logging::log!("inspect_view res_inspect awaited");
                        view! { ... }
                    });
                    count.update_untracked(|x| *x += 1);
                    leptos::logging::log!(
                        "returning result, result.is_ok() = {}, count = {}",
                        result.is_ok(),
                        count.get(),
                    );
                    result
                })
            }}
        </Suspense>
    }                       
```

Where `res_inspect` is the resource wrapping the server function that has a delay (I actually set it to 2.5 seconds) and the following log is what gets generated on the server when navigating directly to `Demo target` from the Home page in CSR:

```
inspect_view closure invoked
inspect_view Suspend::new() called
res_inspect - inspect_item().await  # a 2.5 second pause here
inspect_view closure invoked
inspect_view res_inspect awaited
returning result, result.is_ok() = true, count = 1
inspect_view Suspend::new() called
inspect_view res_inspect awaited
returning result, result.is_ok() = true, count = 2
```

While reloading that to get the SSR rendering, we get this instead:

```
inspect_view closure invoked
inspect_view Suspend::new() called
inspect_view closure invoked
inspect_view Suspend::new() called
res_inspect - inspect_item().await  # the same 2.5 second pause
inspect_view res_inspect awaited
returning result, result.is_ok() = true, count = 1
```

So maybe there is something funny going on here as the SSR rendering is behaving as what OP expected and does the "if there is an .await point that is not ready yet, it will not progress beyond that point" as expected, while CSR isn't doing as per that requirement.

**Comment by gbj** [comment]
@metatoaster This behavior (running 2x on the client) is the opposite of the example in the original issue (running 2x on the server), but sounds more like the bug behavior described (creating websockets 2x).

I hadn't fully thought about this one, so thanks to you both.

What's happening here is:
- `move || Suspend /* ... */` creates a `RenderEffect` like any other `move ||`
- this effect then tracks the resources that are read when you `.await` them, so that if the resources values changes, it will know to render again
- however, resources also notify their dependencies when they resolve, which means that this resource will actually notify the render effect *when it resolves*, causing the effect to re-run, which is why you get the second `Suspend::new()`. If you had additional components in the view, they would run, creating their effects, resources, websockets? etc.

I'll have to think a bit about how to handle this: We definitely do want that `move ||` to subscribe to the resources it reads from, so that it re-runs when they change, but we don't want the fact that they're resolving to run it again, obviously.

**Comment by BakerNet** [comment]
> This behavior (running 2x on the client) is the opposite of the example in the original issue

The original issue included showing the server side double-effect just because I noticed it after the fact.  The real issue for me was always 2x on the client (websockets only created in component on the client).

> If the server function is actually async on the server (i.e., it is not immediately resolved when called) and the side effect is after the .await, you get the result you're anticipating

On the server side, this is true - the result is what I would expect.  But on the client side, everything after the .await is still run twice - and I'd argue this is a bigger deal because:

1. the client side is where components are more likely to have side effects
2. all memory allocated from the first render is going to become garbage collected by WASM GC - so if the component waiting on the resource allocates a lot (mine does) this can cause some extra slowdown.

**Comment by gbj** [comment]
Yes, I just misunderstood the original issue and this is obviously bad.

**Comment by BakerNet** [comment]
FYI - I updated the example branch to include `use_websocket` from leptos-use in case you wanted to use that branch as a testing ground.

https://github.com/user-attachments/assets/b5589499-2ebe-4861-8300-d121ca6f1925

**Comment by gbj** [comment]
I'm starting to feel pretty good about this whole `Suspend` thing! (Famous last words)

#2959 should actually fix this issue: It creates a new reactive observer, which catches all the reactive async reads in the `Suspend`, and then forwards those subscriptions to the outer render effect (if any) *after* the Suspend's async render finishes. I did test this against metatoaster's gist and against the example in this issue, and I believe it should fix the problem without breaking reactivity.

But let me know!

I also want to take the opportunity to introduce a cancellation mechanism here, which was an existing `// TODO` but is fully necessary now, so that if a resource re-triggers the `Suspend` before it's finished loading (i.e., you have two resources and the second one you `.await` changes while the first is still loading) it will cancel the two instead of racing.

**Comment by BakerNet** [comment]
> But let me know!

I will test this tonight and report back.  I'm impressed by your turnaround on these issues!  🙌

**Comment by metatoaster** [comment]
I'll second the rapid turnaround, and I can tell that the CSR is significantly snappier!

However, I will note that in one of the use case in the application I am developing is definitely broken as the reactivity isn't being propagated like it did before.  In short, I implemented portlet components that is rendered on the root component but depends on data from components at the leaf routes; it expect a context through a series of signals and resources to determine what gets rendered (e.g. components can set the context to render links for local navigational purposes). I will investigate this further to see if my implementation can be improved to restore reactivity and I will provide a MVRE should my attempt fail to resolve the issue.

**Comment by metatoaster** [comment]
Actually, there may be a different problem to what I described as I think it may be hydration related and the gist I've provided does demonstrate this issue. The delay of `2500` should be changed to a more reasonable `25`, and I've moved the count `RwSignal` to the `App` component to better illustrate the point ([direct link to current version](https://gist.github.com/metatoaster/2798451bd9bc99d57b73c46d7cc8ca9a)).

From the `Home Page`, go into `Demo target`, then just poke around the `Inspect` link, the `Inspecting` line should change as it should.  Now, _refresh_ the page, the reactivity is broken when navigating back or using the `Inspect` links; this lasts until you leave that component completely by backing out enough or go `Home` straight away and back to visit that component which will restore reactivity.  Note the log files - when the view isn't being reactive, the `res_inspect:` is still invoked, just that the `Suspend` isn't returning the result. Should also note that `count` does not actually increase while the `Suspend` isn't called after hydration.
